modules/training_pipeline/training_pipeline/prompt_templates/prompter.py

```diff
@@ -0,0 +1,107 @@
+"""
+This script defines a PrompterTemplate class that assists in generating 
+conversation/prompt templates. The script facilitates formatting prompts 
+for inference and training by combining various context elements and user inputs.
+
+"""
+
+
+import dataclasses
+from typing import Dict, List, Tuple
+
+
+@dataclasses.dataclass
+class PrompterTemplate:
+    name: str
+    agent_behaviour: str = """
+    >>INTRODUCTION<<
+    {system_message}
+    """
+    roles: List[str] = ("USER", "ASSISTANT")
+
+    def __format_base(self, user_context: str, news_context: str) -> str:
+        """
+        Formats to Falcon compatible template, using `added_tokens`:
+        - INTRODUCTION : specifies agent behaviour
+        - COMMENT : used for context and user description
+        spec: https://huggingface.co/tiiuae/falcon-7b/raw/main/tokenizer.json
+        """
+        base_template = f"""
+        >>COMMENT<<
```

Comment:
I saw that the most common delimiter is `####` in almost all the docs & courses I looked at so far. Is `>>>` something specific to Falcon?

Reply:
Yes, they are specific to `Falcon-7b` tiuuae.
Here's the link to the repo tokenizer: https://huggingface.co/tiiuae/falcon-7b/raw/main/tokenizer.json

All the special tokens used are part of the vocabulary used by the default tokenizer used with falcon 7b.
<img width="365" alt="Captură de ecran din 2023-09-06 la 18 25 21" src="https://github.com/iusztinpaul/hands-on-llms/assets/23241108/8109db5d-f9cc-476d-a377-5899eb33d4a4">

Reply:
Nice. Good job 🔥  I agree. Leave it as it is. But what about renaming the class to `FalconPromptTemplate` as it is something specific to `Falcon`? Maybe we want to add llama support in later iterations, so we can easily differentiate between the two.

Reply:
I agree regarding the FalconPromptTemplate, i was thinking of using `name` as the delimiter between different templates.
We have the base PromptTemplate, and we register different templates to the `template_registry` using the configurable params, such that we can add as many template formats as we want without needing to implement a template class for each of them.

Let me know which way would be a better fresh-start for teaching purposes, I aimed it to be as generic as possible with the current impl.

Reply:
Looking good with the new refactoring. I like it :fire: Even better than my suggestion.

Maybe one last thing is to rename ` sep2:` to `sep_eos` or `sep_eot` ? To know what is the difference between `sep` and `sep2`.